### DISCUSSION TITLE
Prevent Quarto from wrapping LaTeX code in a `table` environment

### DISCUSSION THREAD

**Discussion by vincentarelbundock** [quarto-dev/quarto-cli#9332] [Q&A]
### Description

Hi everyone,

How can I stop Quarto from enclosing LaTeX code in a table environment when using the `tbl-cap` and `label` chunk options?

Background: I maintain [a package to create tables in `R`.](https://vincentarelbundock.github.io/tinytable/) My users want to use `tbl-cap` and `label` chunk options to specify captions and labels in consistent fashion across output formats.

This works well in HTML and markdown formats. For LaTeX, however, Quarto automatically wraps the code in a floating `table` environment. This breaks my package LaTeX code, which relies on the `tabularray` package to produce "long" tables that *cannot* be included in a floating environment like `table`.

I would like to retrieve the value of chunk options and process those internally in my package. The retrieval can be done easily with `knitr::opts_current$get()`. All I am missing is a way to prevent Quarto wrapping.

Thanks!

Example:

````
---
title: "Example"
format: 
   html: default
    pdf:
        keep-tex: true
---

@tbl-iris says...

```{r}
#| tbl-cap: Blah blah
#| label: tbl-iris

library(tinytable)

knitr::opts_current$get("tbl-cap")

knitr::opts_current$get("label")

tt(iris) |> theme_tt("multipage")
```
````

**Comment by mcanouil** [comment]
I don't think you can do that and keep your table being a cross-reference at the same time.
Maybe (did not try) using the fenced div syntax instead of using code cell options.
(Cc @cscheid )

**Reply by vincentarelbundock** [reply]
Thanks for the reply! Do you mean that `@tbl-blahblah` tags won't be converted to `\ref{tbl-blahblah}` in the raw `.tex` file?

**Reply by cscheid** [reply]
> For LaTeX, however, Quarto automatically wraps the code in a floating table environment.

@mcanouil is right. Ultimately, Quarto needs the `table` environment for crossref purposes in LaTeX. I don't know what you're expecting `@tbl-iris` to produce if not a reference to a floating environment of some kind. How are `tabularray` tables referenced in plain latex documents?

**Reply by vincentarelbundock** [reply]
I would like `@tbl-example` to be converted to `\ref{tbl-example}` in the main `.tex` file, as usual.

`tabularray` tables are referenced exactly like tabulars in a normal `table` environment, using `\ref{}`. However, they cannot be inserted in a float, and the labeling is done using environment options `label={}` and `caption={}` rather than the standard `\label{}`.

Example:

```latex
Table \ref{tbl-example} shows that...

\begin{longtblr}[ 
caption = {This is a caption},
label = {tbl-example},
]                     
{                
colspec={Q[]Q[]Q[]Q[]Q[]},
}                  
\toprule
Sepal.Length & Sepal.Width & Petal.Length & Petal.Width & Species \\ \midrule 
5.1 & 3.5 & 1.4 & 0.2 & setosa \\
4.9 & 3.0 & 1.4 & 0.2 & setosa \\
4.7 & 3.2 & 1.3 & 0.2 & setosa \\
\bottomrule
\end{longtblr} 
```

**Reply by vincentarelbundock** [reply]
In other words, the only thing I would like Quarto to do with my table and cross-refs is to convert:

`@tbl-example` -> `\ref{tbl-example}`

And to *not* insert the code in a `\begin{table}` environment. 

I would handle setting the labels and caption in `tabularray` myself, and make sure that `\ref{}` points to a valid LaTeX cross-ref label.

**Reply by mcanouil** [reply]
> @tbl-example -> \ref{tbl-example}

That's Pandoc.

> And to not insert the code in a \begin{table} environment.

That's because you used `label: tbl-*`.
It's a Quarto feature.
Unfortunately, here you want this feature to behave differently on computed content, which I don't think is really doable.
So, I'm afraid, if you want the cross-reference feature and use your R package, you'll have to use conditional content and reusable code cells, to use the feature for non LaTeX format, and handle yourself this for LaTeX format.

**Reply by cderv** [reply]
> > @tbl-example -> \ref{tbl-example}
>
> That's Pandoc.

@mcanouil what do you mean ? I don't think pandoc convert this type of reference - it does only for citation. Quarto is the one outputting raw LaTeX here as part of crossref feature IMO. 

> That's because you used label: tbl-*.

Yes currently `tbl-` prefixed label trigger the cross-reference system, and I don't think we have a way out yet of float environment for tables. 

We need to add support for `tabularray` somehow, or provide an opt-out mechanism for this special processing for advanced use case where a package author would handle the necessary raw LaTeX itself. The former would be best so that it is integrated with the rest of the tables. Otherwise, mixing cross-referenced table won't work well (for numbering and things like that). 

Not sure what is really possible if `tabularray` and `floats` don't work together ... 🤔  Maybe a special treatment for `longtblr` environment is enough. 

For now, it won't work without trying to hack around how cross ref are triggered in Quarto.

**Reply by vincentarelbundock** [reply]
Thanks a lot for taking the time to clarify things, I really appreciate it! Will try to find a workaround.

> Not sure what is really possible if tabularray and floats don't work together ... 🤔 Maybe a special treatment for longtblr environment is enough.

Right. I should have specified that tabularray has three environments (tblr, talltblr, and longtblr). 

What I described above is only a problem for longtblr. So it is admittedly a pretty niche issue...

**Reply by cderv** [reply]
This is the same issue with longtable I guess.  I think we have a special processing for `longtable`, and this may need to be extended to `longtblr` then 🤔 

* https://github.com/quarto-dev/quarto-cli/issues/7604
* https://github.com/quarto-dev/quarto-cli/pull/7613

**Reply by vincentarelbundock** [reply]
Right. The tricky thing is that IIRC, `longtable` uses the more standard `\caption{}` format. Figuring out where to insert `caption={}` in `longtblr` might not be as straigthforward.

**Comment by andrewheiss** [comment]
Another use case/question for using `tabularray` with Quarto. Right now, because of the issues discussed above by @vincentarelbundock and everyone else, I'm having to do this for multiple-output documents:

````
---
title: "Table stuff"
format: 
  pdf: default
  html: default
---

```{r}
library(tinytable)

tmp <- rbind(mtcars, mtcars)[, 1:6]
```

::: {.content-visible when-format="pdf"}

\ref{tbl-long-thing} will only appear in PDF

```{r}
tt(tmp, width = 0.8, caption = "A long table \\label{tbl-long-thing}") |>
  theme_tt("multipage", rowhead = 1)
```

:::

::: {.content-visible when-format="html"}

@tbl-long-thing will only appear in HTML

```{r}
#| label: tbl-long-thing
#| tbl-cap: "A long table"

tt(tmp, width = 0.8)
```

:::

````

Is that the approach we should be using until the dust settles on `tabularray` and Quarto? Or is there a better way? 

I experimented with conditional chunk options and conditional cross reference container divs:

```
:::{ #tbl-long-thing .content-visible ...}
...
:::
```

but those don't seem to work :(
